In [53]:
import pandas as pd

import nltk
from nltk.corpus import stopwords
from nltk import word_tokenize
from nltk.stem import PorterStemmer
from nltk.stem import WordNetLemmatizer
from nltk.corpus import wordnet

import string
import re

from sklearn.feature_extraction.text import CountVectorizer
from sklearn import svm

from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfVectorizer


from sklearn.naive_bayes import MultinomialNB
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV

from sklearn.model_selection import train_test_split
from sklearn.metrics import plot_confusion_matrix
from sklearn.metrics import classification_report

from sklearn import tree

from sklearn import metrics
from sklearn.metrics import classification_report

from sklearn.metrics import accuracy_score, precision_score, recall_score, confusion_matrix, f1_score, roc_auc_score 

nltk.download('stopwords')
nltk.download('punkt')
nltk.download('wordnet')
stopwords = stopwords.words('english')

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\13473\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\13473\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\13473\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


In [2]:
df = pd.read_csv('spam_ham_dataset.csv')
df.head()

Unnamed: 0 label                                               text  \
0         605   ham  Subject: enron methanol ; meter # : 988291\r\n...   
1        2349   ham  Subject: hpl nom for january 9 , 2001\r\n( see...   
2        3624   ham  Subject: neon retreat\r\nho ho ho , we ' re ar...   
3        4685  spam  Subject: photoshop , windows , office . cheap ...   
4        2030   ham  Subject: re : indian springs\r\nthis deal is t...   

   label_num  
0          0  
1          0  
2          0  
3          1  
4          0

In [3]:
print(df.isnull().sum())
print(df.duplicated().sum())

Unnamed: 0    0
label         0
text          0
label_num     0
dtype: int64
0


In [4]:
def make_lower(a_string):
    return a_string.lower()

def remove_punctuation(a_string):    
    a_string = re.sub(r'[^\w\s]','',a_string)
    return a_string

def remove_stopwords(a_string):
    words = word_tokenize(a_string)
    valid_words = []
    for word in words:
        if word not in stopwords:
            valid_words.append(word)
    a_string = ' '.join(valid_words)
    return a_string

def text_pipeline(input_string):
    input_string = make_lower(input_string)
    input_string = remove_punctuation(input_string)
    input_string = remove_stopwords(input_string)    
    return input_string

In [5]:
df['text_clean'] = df['text'].apply(text_pipeline)

## Data Dictionary

| Column      | Data Type   | Description | 
| :-----------: | :-----------: | :----------- |
| Label       | String      | labeled as spam or ham representing if the text is spam or not |
| Text        | String      | The emails |
| Label_num   | Int         | A 0 or 1 representing spam or not respectively |
| Text_clean  | String      | Cleaned version of the emails after passed through the text pipeline |

In [6]:
X = df['text_clean'].values

y = df['label_num'].values

In [30]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42)

In [31]:
cv = CountVectorizer()
features = cv.fit_transform(X_train)

In [32]:
model = svm.SVC()
model.fit(features,y_train)

SVC()

In [34]:
# SVC
features_test = cv.transform(X_test)
y_pred = model.predict(features_test)

accuracy = accuracy_score(y_true=y_test, y_pred=y_pred)
print("Accuracy Score: %f" % accuracy)

precision = precision_score(y_true=y_test, y_pred=y_pred)
print("Precision Score: %f" % precision)

recall = recall_score(y_true=y_test, y_pred=y_pred)
print("Recall Score: %f" % recall)

f1 = f1_score(y_true=y_test, y_pred=y_pred)
print('F1 Score: %f' % f1)

Accuracy Score: 0.958454
Precision Score: 0.890625
Recall Score: 0.972696
F1 Score: 0.929853


In [35]:
model.get_params()

{'C': 1.0,
 'break_ties': False,
 'cache_size': 200,
 'class_weight': None,
 'coef0': 0.0,
 'decision_function_shape': 'ovr',
 'degree': 3,
 'gamma': 'scale',
 'kernel': 'rbf',
 'max_iter': -1,
 'probability': False,
 'random_state': None,
 'shrinking': True,
 'tol': 0.001,
 'verbose': False}

In [65]:
cv = CountVectorizer()
features = cv.fit_transform(X_train)
features_test = cv.transform(X_test)


In [66]:
## RandomForestClassifier
random_forest_model = RandomForestClassifier()

random_forest_model.fit(features, y_train)

y_pred = random_forest_model.predict(features_test)

accuracy = accuracy_score(y_true=y_test, y_pred=y_pred)
print("Accuracy Score: %f" % accuracy)

precision = precision_score(y_true=y_test, y_pred=y_pred)
print("Precision Score: %f" % precision)

recall = recall_score(y_true=y_test, y_pred=y_pred)
print("Recall Score: %f" % recall)

f1 = f1_score(y_true=y_test, y_pred=y_pred)
print('F1 Score: %f' % f1)

Accuracy Score: 0.974879
Precision Score: 0.946488
Recall Score: 0.965870
F1 Score: 0.956081


In [70]:
random_forest_model.get_params()

{'bootstrap': True,
 'ccp_alpha': 0.0,
 'class_weight': None,
 'criterion': 'gini',
 'max_depth': None,
 'max_features': 'auto',
 'max_leaf_nodes': None,
 'max_samples': None,
 'min_impurity_decrease': 0.0,
 'min_impurity_split': None,
 'min_samples_leaf': 1,
 'min_samples_split': 2,
 'min_weight_fraction_leaf': 0.0,
 'n_estimators': 100,
 'n_jobs': None,
 'oob_score': False,
 'random_state': None,
 'verbose': 0,
 'warm_start': False}

## Exploratory Analysis

The problem we are trying to address is spam email detection. We started by cleaning the emails. This included lower casing the email, removing punctuation and removing stop words. We then split our data into a testing and traning set. We then passed our training data into a count vectorizer which converts a collection of text documents to a matrix of token counts. Essentailly it counts the occurance of each word and assigns a token to that word. This allows use to pass words into a model which is what we need. We used 2 different models, Support Vector Classifier (SVC) and RandomForestClassifer. Lastly we measured the performace of both the models. Both models performed well, but the random forest classifer had slightly better performance with an accuracy score of 0.97, precision score of 0.94, recall Score of 0.96, and f1 score of 0.95.


## Challenges

We knew from the beginning this was a classification problem and so we knew what types of models we needed, however the challenge was the nlp part. We needed to somehow input text to a model. This is where we discovered count vectorization, which allowed us to pass the text into the models. A current challenge we have is getting further insights of the model. The models are performing well, but they don't give us any insightful information on spam emails. We need to futher investigate what exactly we are doing and extrapolate any further information.

## Future Work

For the future we want to test different hyperparameters to determine the best ones so that we can have the best model possible. We also want to further investigate how the model is making its desicion. We want to see if there are any specific words that most spam emails have or if there are other identifiers. Lastly we want to try to further our investigation. We want to ask ourselves better questions and gain further insights on the problem of spam emails.

## Contributions

We contributed equalling. We were on a call, peer programming.